### ISSUE TITLE
Optimizer transforms nested joins incorrectly

### ISSUE THREAD

**Issue by 10bas10** [tobymao/sqlglot#1898]
Hey @GeorgeSittas, thanks a lot for the fix on the nested joins. That will work on my part, as i only need to be able to parse and use the scope class. However i did a check on the optimizer as well and the SQL seems to be transformed into SQL that is functionally different from the original.

As you can see in the output the query is beeing transformed into regular joins and the table b and c arent nested anymore.

The output:
```
SELECT
  "a"."id" AS "id",
  "b"."test" AS "test",
  "c"."test2" AS "test2"
FROM "a" AS "a"
LEFT JOIN "b" AS "b"
  ON "b"."id" = "a"."id"
JOIN "c" AS "c"
  ON "c"."id" = "b"."id"
```

The code i used:
```
import sqlglot
from sqlglot import optimizer
expr = sqlglot.parse_one("""
SELECT
 a.id
,b.test
,c.test2
FROM a
LEFT JOIN (
    b
    INNER JOIN c
        ON c.id = b.id
)
    ON b.id = a.id
""", read="tsql")

schema={
    "testcat": {
        "testdb": {
            "a": {
                "id": "INT",
            },
            "b": {
                "id": "INT",
                "test": "STRING",
            },
            "c": {
                "id": "INT",
                "test2": "STRING",
            },
        }
    }
}

expr = optimizer.optimize(
    expression=expr,
    schema=schema,
)

postgres = sqlglot.Dialect.get_or_raise("postgres")()
print(postgres.generate(expr, pretty=True))
```

**Comment by GeorgeSittas** [comment]
Hey @10bas10, happy to know your issue was resolved.

> However i did a check on the optimizer as well and the SQL seems to be transformed into SQL that is functionally different from the original.

So after discussing this, we agreed on flattening the `JOIN` sequence in the optimizer because preserving the parentheses wasn't a trivial thing to do. This shouldn't actually be "functionally different" (as in "different semantics"), because AFAIK specifying an explicit `JOIN` order is simply an optimization and the parentheses don't change the returned row set. This is behaviour is now documented in the docstring of `qualify_tables`.

```
    Note:
        This rule effectively enforces a left-to-right join order, since all joins
        are unnested. This means that the optimizer doesn't necessarily preserve the
        original join order, e.g. when parentheses are used to specify it explicitly.
```

Note that this is consistent with SQLGlot's optimizer in general: it doesn't make any guarantees about the JOIN order. Even without my recent PR, you could get different `JOIN` orders for some optimized queries.

**Comment by GeorgeSittas** [comment]
I'm gonna go ahead to close this one, but happy to continue the discussion either here or in Slack.

**Comment by 10bas10** [comment]
Hey @GeorgeSittas,

It actually is functionally different. Let me demonstrate a case where the results in differerent output, it is actually the same case as i provided in this bug.

In the following query, where nested joins are applied, the inner join is applied in a different context level. The inner join is applied to the table b.

As you can see this query gives all records that are in table a even with the joins not supplying any matches:
![image](https://github.com/tobymao/sqlglot/assets/81879308/3aca1be2-1f12-4b67-9d96-6ce5736fcfc9)

```
;WITH a AS (
	SELECT 1 AS id
),
b AS (
	SELECT 2 AS id, 'b' as test
),
c AS (
	SELECT 2 AS id, 'c' as test2
)
SELECT
 a.id
,b.test
,c.test2
FROM a
LEFT JOIN (
    b
    INNER JOIN c
        ON c.id = b.id
)
    ON b.id = a.id
```

In the query down below, actually the query beeing outputted by the optimizer, the join is indeed flattened. However this gives a different end result as the inner join is actually performed in a different context. This query doesnt give any output as you can see:
![image](https://github.com/tobymao/sqlglot/assets/81879308/b484ad69-819a-4d99-bf8d-5f98cd87b6a6)


```
;WITH a AS (
	SELECT 1 AS id
),
b AS (
	SELECT 2 AS id, 'b' as test
),
c AS (
	SELECT 2 AS id, 'c' as test2
)
SELECT
 a.id
,b.test
,c.test2
FROM a
LEFT JOIN b
	ON b.id = a.id
INNER JOIN c
	ON c.id = b.id
```

**Comment by GeorgeSittas** [comment]
I see, thanks for the explanation -- it seems like this is indeed incorrect then. I'll try to resolve this.

**Comment by GeorgeSittas** [comment]
@10bas10 I've fixed your issue in #1900 and am currently working on improving coverage.

Thanks for reporting this problem, hopefully the optimizer will now be a little more robust!
